### PR TITLE
first cut at jit device placement api

### PR DESCRIPTION
**Pull request by mattjj**
cf. #1009

The idea is that we want to associate compiled computations (both `pmap` and `jit` ones, though this PR is just starting with `jit`) with physical device numbers so that we can control where things are executed (and as a consequence where the results are produced and lazily persisted). The API here is just to provide an optional `device_assignment` argument to `jit` that closely follows [the XLA DeviceAssignment API](https://github.com/tensorflow/tensorflow/blob/2522738eba2be57e524b48ad40c15f88f5464b3f/tensorflow/compiler/xla/python/xla_client.py#L431-L449). In particular, to assign a computation to physical device 1 we could write something like

```python
@partial(jit, device_assignment=1)
def foo(x, y, z):
  ...
```

### CODE REVIEW

**Comment by christopherhesse** [comment]
Is there a way to associate non-compiled computations/jp.array() calls to a particular device?

**Comment by mattjj** [comment]
@christopherhesse That's a good idea. For `jp.array` calls, maybe you can use `jax.device_put` with its `device_num` argument?

For non-jitted computations, i.e. what we call op-by-op stuff, would a context manager make sense?

**Comment by cshesse** [comment]
A context manager is what `cupy` uses, and would make sense to me.

**Comment by skye** [comment]
I'm not sure if this is a good idea, but what about a function decorator instead of a context manager?

In my experience with TF, it can be hard to debug issues with context managers because you have to manually look at each stack frame and see if you're in a relevant context (or use a debugger to inspect the global state set by the context manager). A function decorator would have the advantage of being immediately visible in a stack trace. OTOH maybe it'll be more unwieldy to use... What do people think?

**Comment by mattjj** [comment]
Hmm, in a way jit is already the function decorator version. 

We could also have the policy that in op-by-op operation the computation runs wherever the arguments live. If it’s ambiguous then the choice could be arbitrary, so then users could ensure things run on the correct device by calling device_put on the inputs. 

Would that replace the need for a context manager or new decorator for op-by-op stuff?

**Comment by skye** [comment]
I like this idea! We could potentially extend this to jit and pmap too. The only downside I see is that it's less explicit, i.e. maybe there's the potential for unpredictable performance cliffs depending on what kind of input you get. But I think it's still better than an explicit context or decorator. I think pytorch may work this way?

**Comment by hawkinsp** [comment]
What happens for nullary operators, e.g., `np.arange`? We still need a mechanism for placing those.

The thing I like about the decorator approach is that it's consistent with the rest of the system: use same tracing mechanism for everything.

**Comment by mattjj** [comment]
Well, the context manager will help with nullary operators, but tracing won't: tracing is the thing that requires data dependence, after all.

I think explicit `device_put` calls actually might work well with nullary operators: just call it on the output! Wherever we do something sensible with those (e.g. in this case the output of `np.arange` might be one of our lazy iota expressions, and so we'd hope to materialize it on the device directly, or perhaps delay further and just propagate a lazy constant with a device id tag) then we'd do something sensible here too.

Of the three solutions for op-by-op (tracing, context manager, and explicit `device_put` calls) I currently prefer the latter, just based on the discussion here so far. Other points in favor / against?

**Comment by mattjj** [comment]
I experimented a bit with the op-by-op stuff but I decided I'd rather leave that for a follow-up PR. Dealing with `pmap` is  also for a follow-up.

I added a commit (and rebased things) that makes `_execute_compiled` place the arguments on the correct device for the computation. I think this change, minimal as it is (just concerning `jit`) is ready for review.

I might tweak the API so that the `device_assignment` arg to `jit` takes an int rather than a tuple.

File: jax/api.py [hunk 1]
```diff
@@ -90,6 +90,9 @@ def jit(fun, static_argnums=()):
       different values for these constants will trigger recompilation. If the
       jitted function is called with fewer positional arguments than indicated
       by `static_argnums` then an error is raised. Defaults to ().
+    device_assignment: Optional, an int specifying the device ordinal for which
```
Comment (hawkinsp) [inline jax/api.py]: Just a caution: I think it highly likely that we will stop using ints to name devices here soon (e.g., it doesn't necessarily work that well in a multihost setup). I'm not sure whether you wish to add a caveat to the documentation that this API is not yet stable.

**Review by hawkinsp — APPROVED** [review]
This seems fine as a first cut. My main concern is I would like to let the API see some use before we declare it baked, so suitable caveats in the documentation might be wise.

File: jax/api.py (hunk 1 above)
Comment (mattjj) [reply jax/api.py]: Great point!

**Comment by jekbradbury** [comment]
I agree that this is a good first version! Just to have it written down somewhere, these are my thoughts from last week about one possibility of where to go from here/how to extend this to pmap and multihost:

- Each pmap could take a replica placement thing (function, dict, idk) that maps devices (whether those are integers or something else is orthogonal) to replica IDs. So, for instance, "map the TPU mesh x axis to the replica axis" is a valid replica placement, as is "map GPUs to replica IDs based on which host they're connected to"; it should probably be valid to map only a subset of available devices.
- These partitions of available devices can then compose in the nested-pmap context: if you have a pmap partitioning devices based on cores on each host, and another pmap partitioning based on hosts, then e.g. collectives that use the axis from the first pmap will act collectively on each host and run a pure map between the hosts while collectives that use both axes will act collectively over all active devices.
- The number of devices to activate is the total number of partitions created by composing the various nested replica placement partitions (i.e., if you compose two identical 4-way partitions you'll activate 4 devices; if you compose two orthogonal 4-way partitions you'll activate 16 devices).
